### ISSUE TITLE
L034 unexpected complexity order

### ISSUE THREAD

**Issue by ixti** [sqlfluff/sqlfluff#4171]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

I might be wrong, but `some_function(...)::some_type` should have higher complexity than `some_function(...)`.

### Expected Behaviour

I expect the following SQL to be valid:

``` sql
SELECT
    concat('1970-01-01', ' ', '12:00:00') AS a,
    concat('1970-01-01', ' ', '12:00:00')::date AS b
```

### Observed Behaviour

SQLFluff says it should be:

``` sql
SELECT
    concat('1970-01-01', ' ', '12:00:00')::date AS b,
    concat('1970-01-01', ' ', '12:00:00') AS a
```

### How to reproduce

``` bash
sqlfluff lint --dialect postgres --verbose - <<SQL 
SELECT
    concat('1970-01-01', ' ', '12:00:00') AS a,
    concat('1970-01-01', ' ', '12:00:00')::date AS b
SQL
```

### Dialect

postgres
snowflake

### Version

sqlfluff, version 1.4.4
Python 3.10.9

### Configuration

``` ini
[sqlfluff]
templater = jinja
```

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by peterbud** [comment]
With the latest sqlfluff the above statement is valid, and is NOT marked as a linting error.
